src/v/kafka/server/handlers/metadata.cc

```diff
@@ -157,7 +158,24 @@ metadata_response::topic make_topic_response_from_topic_metadata(
         auto lt = get_leader_term(tp_ns, p_as.id, md_cache, replicas);
         if (lt && !is_node_isolated && p.error_code == error_code::none) {
             p.leader_id = lt->leader.value_or(no_leader);
-            p.leader_epoch = leader_epoch_from_term(lt->term);
+
+            // If we don't have term information for the given partition,
+            // submit a stale guess. If we are wrong, the client is expected to
+            // fence this and eventually attempt to get metadata again.
+            p.leader_epoch = leader_epoch_from_term(
```

Comment:
Lgtm but the logic looks weird and will be easy to mess it up in the future.

I would prefer this to be explicit, something like:

```cpp
if (lt->term.has_value()) {
  p.leader_epoch = ... lt->term;
} else {
  // explanation for this branch
  p.leader_epoch = ... 0;
  p.error_code = error_code::leader_not_available;
}
```